### DISCUSSION TITLE
Publish Individual Locales

### DISCUSSION THREAD

**Discussion by zubricks** [payloadcms/payload#4194] [Roadmap]
Provide the ability for users to publish one locale at a time e.g. promote only Spanish, German, etc. to prod.

Feature breakdown:

1. Publish button will be extended to include a dropdown selector, this will display all available locales to publish 
2. When publishing a single locale, only the data for the selected locale will be updated on the published doc

This is applicable to collections and globals.

Estimated Date of Completion: **8th Dec**

**Comment by tylandavis** [comment]
## Design Intent

We'll be updating the button component to allow for a dropdown of actions. Clicking on the main part of the button will publish the full document's changes, while clicking on 'Publish [Locale] only' will only publish changes from that locale's data, and keep other locale's changes in their draft state.
 
![new-publish-button](https://github.com/payloadcms/payload/assets/89618855/5525a9eb-0419-4c4d-bd74-0919ffedcede)

**Reply by PP-Tom** [reply]
I think publish should publish the current local and the dropdown could say "Publish All", I think that would be a editor experience. The example seems too easy to accidentally publish all.

**Comment by Janisvb** [comment]
Is there an updated timeline for this feature?

**Comment by jhnsml** [comment]
Any update here?

**Comment by Lismont** [comment]
also interested in this feature 👋

**Comment by tylandavis** [comment]
Hey everyone, we're all deep into getting 3.0 beta out, and should be ready to release in the next week or so. Once that's complete we will be back to working on roadmap features, and this one is definitely one of the top priorities for us.

**Comment by felixhagspiel** [comment]
@tylandavis would it be possible for me to implement a temporary workaround using the `beforeChangeHook` until you release v3? I.e. is it possible to tell the backend to just update the currently selected locale? I have not found any info on this in the docs

**Comment by PP-Tom** [comment]
+1 for this feature, it's really important for a project I have at the moment. Looking for alternative implantation for now, but hitting road blocks.

**Comment by johannesspohr** [comment]
I had a look at the [ongoing implementation](https://github.com/payloadcms/payload/pull/7669).
First, I'm really excited that we're getting this soon, thanks for the work on that!

I tried it for a bit and want to second what @PP-Tom mentioned on the design. If I'm working on a specific locale, the only two use cases I see are publishing the current locale and publishing all locales, but not any other locale than the current one. So my expectation is that the default action is publishing the current locale, and in the dropdown menu there would be an option to publish all locales. 
Also, if I understood it correctly, it's not possible at the moment that one locale is in draft state, while another one is published (say an article is available in one locale, but not in another). I guess you could solve that with some hooks magic and an additional field, but to me it seems a bit counterintuitive that publishing one locale also sets another locale to published state without actually publishing the contents.

**Comment by ars3nicum** [comment]
+1

**Comment by denolfe** [comment]
Released in 3.0

Relevant [PR](https://github.com/payloadcms/payload/pull/7669)

**Comment by cbratschi** [comment]
The locale dropdown is not fully visible  here. Probably a z-index issue.

Is there an API to get the list of published locales of a document? I checked the source code and Payload creates snapshots in the revision collection.

Still migrating our code to Payload 3.x and this is soon next on our migration list.

**Reply by tylandavis** [reply]
@cbratschi can you open an issue with screenshots of the issue you are running into?

**Reply by cbratschi** [reply]
Created here: https://github.com/payloadcms/payload/issues/9699

**Reply by kevmiota** [reply]
@cbratschi have you found a way to retrieve published locales through the api?

**Reply by cbratschi** [reply]
@kevmiota yes, see the used code below: https://github.com/payloadcms/payload/discussions/4194#discussioncomment-11586495. All published and draft locales are available in each document. We use this to check if a locale is available and all required fields are present.

**Comment by cbratschi** [comment]
I reviewed this new Payload 3 feature and in my opinion the handling is confusing and I doubt my customers would understand the currently implemented behavior and easily wrong content gets published.

The current behavior is:

- can edit multiple locales in parallel (fine)
  - cannot see which ones are modified (bad)
- can publish a single locale of those locales (fine)
  - why has the dropdown only the active locale? Have to switch to another locale and the use the dropdown again to publish that one.
- can unpublish all published locales (due to global _status per document)
  - cannot unpublish a single locale (bad)
  - after unpublish publish dropdown can select to publish a single locale but it publishes all previously unpublished locales (bug or limitation?)

We would like to have:

- Fields:
  - _status per language (published, draft)
  - createdAt per language (first language version)
  - updatedAt per language (last language edit)
  - publishedAt per language (first published date)
    - Can be in future to use publish task.
- UI:
  - Publish/unpublished a single language version.
  - Publish/unpublished all available language versions.
- Easy way to get the _status of all languages (e.g. to create language switch components on frontend)

Some of those features could be implemented on top of Payload in custom fields but in my opinion those are core features which should be available to all users.

**Comment by cbratschi** [comment]
Wrote an initial version of a locales plugin. It basically works but if more than one locale is published, it's currently impossible to know which ones got published therefore the status value is wrong. The beforeChangeHook fires only once.

```TypeScript
import type { CollectionConfig, Config, Field, GlobalConfig, GroupField, CollectionBeforeChangeHook, PayloadRequest } from 'payload';

//flags
const DEBUG = false;

/**
 * Locale plugin configuration.
 */
export interface LocalePluginConfig {
    //no params
}

/**
 * Locale data of the document.
 */
type DocumentLocales = Record<string, DocumentLocale>;

/**
 * Locale info.
 */
interface DocumentLocale {
    /**
     * Locale status.
     */
    status: 'draft' | 'published'

    /**
     * Creation date of the locale.
     */
    createdAt: Date

    /**
     * Update date of locale.
     */
    updatedAt: Date

    /**
     * Published date of the locale.
     */
    publishedAt?: Date
}

/**
 * Stores locale data in a locales group field.
 *
 * @param _pluginConfig
 * @returns
 */
export default function localePlugin(_pluginConfig: LocalePluginConfig) {
    //modify configuration
    return (config: Config) => {
        //modify collections
        const collections = (config.collections || []).map(collection => {
            //create new collection data
            return <CollectionConfig>{
                ...collection,

                hooks: {
                    ...(collection.hooks || {}),

                    beforeChange: [
                        ...(collection.hooks?.beforeChange || []),

                        //collect locale data
                        ({ data, operation, req }) => handleBeforeChange(collection.slug, data, operation, req)
                    ]
                },

                fields: [
                    ...(collection.fields || []),

                    createLocalesGroup(config)
                ]
            };
        });

        //modify globals
        const globals = (config.globals || []).map(global => {
            //create new collection data
            return <GlobalConfig>{
                ...global,

                hooks: {
                    ...(global.hooks || {}),

                    beforeChange: [
                        ...(global.hooks?.beforeChange || []),

                        //collect locale data
                        ({ data, req }) => handleBeforeChange(global.slug, data, 'update', req)
                    ]
                }
            };
        });

        return <Config>{
            ...config,

            collections,
            globals
        };
    };
}

/**
 * Handle before change.
 *
 * @param slug
 * @param data
 * @param operation
 * @param req
 * @returns
 */
function handleBeforeChange(slug: string, data: Partial<any>, operation: Parameters<CollectionBeforeChangeHook>[0]['operation'], req: PayloadRequest) {
    //check locale is set
    const locale = req.locale;

    if (!locale) {
        //log
        req.payload.logger.error('Locale missing!');

        return data;
    }

    //update fields
    const localesData: DocumentLocales = data.locales || {};
    const localeData: DocumentLocale = localesData[locale] ?? {};
    const now = new Date();

    localesData[locale] = localeData;

    // 1) status (keep global value)
    //cbxx FIXME fails if two languages are published -> only first one set as published!!!
    //cbxx 1) create new document 2) edit two languages, 3) publish both at once -> single call with the currently active language
    localeData.status = data._status;

    // 2) creation date (store once using current date)
    if (!localeData.createdAt) {
        localeData.createdAt = now;
    }

    // 3) update date (now)
    localeData.updatedAt = now;

    // 4) publishedAt (first publish)
    if (operation === 'create' || operation === 'update') {
        if (data._status === 'published' && !localeData.publishedAt) {
            localeData.publishedAt = now;
        }
    }

    if (DEBUG) {
        console.log(`Locales ${slug} ${data.id ?? ''}:`);
        console.dir(localesData);
    }

    data.locales = localesData;

    return data;
}

/**
 * Create the locales group field.
 *
 * @param config
 * @returns
 */
function createLocalesGroup(config: Config): GroupField {
    return                     {
        //locales
        name: 'locales',
        type: 'group',

        admin: {
            components: {
                //no edit
                Field: false
            }
        },

        fields: createLocaleFields(config)
    };
}

/**
 * Get locale group and its fields.
 *
 * @param config
 * @returns
 */
function createLocaleFields(config: Config): Field[] {
    if (!config.localization) {
        return [];
    }

    return config.localization.locales.map(locale => {
        const name = typeof locale === 'string' ? locale:locale.code;

        return <GroupField>{
            //locale group
            name,
            type: 'group',

            fields: [{
                //status
                name: 'status',
                type: 'select',

                options: [{
                    label: 'Draft',
                    value: 'draft'
                }, {
                    label: 'Published',
                    value: 'published'
                }]
            }, {
                //createdAt
                name: 'createdAt',
                type: 'date'
            }, {
                //updatedAt
                name: 'updatedAt',
                type: 'date'
            }, {
                //publishedAt
                name: 'publishedAt',
                type: 'date',

                admin: {
                    date: {
                        pickerAppearance: 'dayAndTime'
                    },

                    position: 'sidebar'
                }
            }]
        };
    });
}

//cbxx TODO scheduled published (date range)
```

**Comment by maximseshuk** [comment]
I have the same problem. I want to be able to publish in just one specific language. Otherwise, it loses its purpose because it's available in both languages anyway, even if the translation isn't finished.

**Reply by baonhat27** [reply]
Any update on this ?

**Reply by tkrotoff** [reply]
I believe this has been implemented here: https://github.com/payloadcms/payload/pull/7669
and is available since [v3.0.0-beta.108](https://github.com/payloadcms/payload/releases/tag/v3.0.0-beta.108)

**Reply by rktyt** [reply]
I don't think the implementation is sufficient, but for now, it seems like the only way to retrieve it is by using the method mentioned in https://github.com/payloadcms/payload/discussions/5469#discussioncomment-8985636.
